src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,474 @@
+        try {
+            Segment requestQueue = requestSegments
+                .computeIfAbsent(
+                    SegmentPriority.MEDIUM == request.getPriority() ? request.getDetectorId() : request.getPriority().name(),
+                    k -> new Segment()
+                );
+
+            requestQueue.setLastAccessTime(clock.instant());
+            requestQueue.put(request);
+        } catch (Exception e) {
+            LOG.error(new ParameterizedMessage("Failed to add requests to [{}]", this.queueName), e);
+        }
+    }
+
+    private void maintainForThreadPool() {
+        for (final ThreadPoolStats.Stats stats : threadPool.stats()) {
+            String name = stats.getName();
+            // cold entity queue mostly use these 3 threadpools
+            if (ThreadPool.Names.SEARCH.equals(name) || ThreadPool.Names.GET.equals(name) || ThreadPool.Names.WRITE.equals(name)) {
+                if (stats.getQueue() > (int) (maxQueuedTaskRatio * threadPool.info(name).getQueueSize().singles())) {
+                    setCoolDownStart();
+                    break;
+                }
+            }
+        }
+    }
+
+    private void prune(Map<String, Segment> segments) {
+        for (Map.Entry<String, Segment> segmentEntry : segments.entrySet()) {
+            if (segmentEntry.getKey().equals(SegmentPriority.HIGH.name())) {
+                return;
+            }
+            // remove more requests in the low priority segment
+            float removeRatio = mediumSegmentPruneRatio;
+            if (segmentEntry.getKey().equals(SegmentPriority.LOW.name())) {
+                removeRatio = lowSegmentPruneRatio;
+            }
+
+            Segment segment = segmentEntry.getValue();
+
+            if (segment == null) {
+                return;
+            }
+
+            BlockingQueue<RequestType> segmentContent = segment.content;
+            // remove 10% of old requests
+            int deletedRequests = (int) (segmentContent.size() * removeRatio);
+            while (false == segmentContent.isEmpty() && deletedRequests-- >= 0) {
+                segmentContent.poll();
+            }
+        }
+    }
+
+    private void maintainForMemory() {
+        // removed expired segment
+        maintenance(requestSegments, stateTtl);
+
+        if (isSizeExceeded()) {
+            // remove until reaching below queueSize
+            do {
+                prune(requestSegments);
+            } while (isSizeExceeded());
+        } else if (adCircuitBreakerService.isOpen()) {
+            // remove a few items in each segment
+            prune(requestSegments);
+        }
+    }
+
+    private boolean isSizeExceeded() {
+        Collection<Segment> segments = requestSegments.values();
+        int totalSize = 0;
+
+        // When faced with a backlog beyond the limit, we prefer fresh requests
+        // and throws away old requests.
+        // release space so that put won't block
+        for (Segment segment : segments) {
```

Comment:
Can this be simplified with a stream?

Reply:
Stream has its own problem: https://stackoverflow.com/questions/44180101/in-java-what-are-the-advantages-of-streams-over-loops